Color percentage indicator

MarcoGorelli opened the pull request:
Fixes https://github.com/pytest-dev/pytest/issues/6097.

MarcoGorelli commented:
I'm aware I'm supposed to write tests before writing code, I just haven't (yet) figured out how to test this, sorry

blueyed commented:
No worries, if you're lucky you can adjust an existing one.
(I've edited the original comment, to add a reference, and remove the boilerplate)
Thanks for working on this! :)

blueyed commented:
While working on this we should keep in mind to use it also for the "SKIPPED", "FAILED", "XFAIL", etc prefixes in the short test summary.

blueyed commented:
CI passes (except for the flaky hypothesis issue), so you need a new test.
I suggest using one based on `monkeypatch.setenv("PY_COLORS", "1")`, so that color output is forced.

MarcoGorelli commented:
Thanks, I was wondering about the hypothesis issue.

OK, I'll give that a go and will let you know

blueyed commented on testing/test_terminal.py:
```diff
@@ -1835,13 +1835,13 @@ def test_collecterror(testdir):
+            "*! Interrupted: 1 error during collection !*",
             "*= 1 error in *",
         ]
     )
```
Yes, bad/outdated merge.
Cherry-picked into features in 9303de8.

MarcoGorelli commented:
I've added tests with `monkeypatch.setenv("PY_COLORS", "1")`, is this what you had in mind?

blueyed commented:
> I've added tests with `monkeypatch.setenv("PY_COLORS", "1")`, is this what you had in mind?

Yes.  Although I think we should them more readable (via format strings for color escapes), so that it would use `"{red}"` there then.
Also I am not sure if the base/default test is really needed, and if it would be faster to have them all in a single run (have not looked at what the "many files" fixture really does).
Looking forward to testing this out later..!

MarcoGorelli commented:
Ok, I've removed two of the tests and have kept the one that has all three colors in it (green, red, yellow). I've also added constants to make the tests more readable.

Just realised that pytest-CI checks for Python 3.5 to work too, so no f-strings....I'll change this later today

blueyed commented:
Nice!  You also need a changelog then (see `changelog/README.rst`).

MarcoGorelli commented:
Thanks! Changelog added

blueyed commented:
Thanks, will review it properly later - you could however squash and force-push it already (sorry, I'd like we could just do it here (#4361)).

blueyed commented:
Awesome!
